projects/snapshot_bot/config.cc

```diff
@@ -0,0 +1,271 @@
+        cv::read(node["shouldSaveTestingDiagnostic"], m_SaveTestingDiagnostic, m_SaveTestingDiagnostic);
+        cv::read(node["shouldStreamOutput"], m_StreamOutput, m_StreamOutput);
+        cv::read(node["shouldUseODK2"], m_ODK2, m_ODK2);
+        cv::read(node["shouldRecordVideo"], m_RecordVideo, m_RecordVideo);
+        cv::read(node["videoCodec"], m_VideoCodec, m_VideoCodec);
+        cv::read(node["videoFileExtension"], m_VideoFileExtension, m_VideoFileExtension);
+
+        // Assert that configuration is valid
+        BOB_ASSERT(!m_UseBinaryImage || !m_UseHorizonVector);
+
+        // **YUCK** why does OpenCV (at least my version) not have a cv::read overload for std::string!?
+        cv::String outputPath;
+        cv::read(node["outputPath"], outputPath, m_OutputPath.str());
+        if (outputPath.empty()) {
+            m_OutputPath = filesystem::current_path() / "training";
+        } else {
+            m_OutputPath = (std::string)outputPath;
+        }
+
+        cv::read(node["maxSnapshotRotateDegrees"], m_MaxSnapshotRotate, m_MaxSnapshotRotate);
+
+        readIntegerSize(node["pmFwdLASize"], m_PMFwdLASize, m_PMFwdLASize);
+        readIntegerSize(node["pmFwdLAIncreaseSize"], m_PMFwdConfig.increaseSize, m_PMFwdConfig.increaseSize);
+        readIntegerSize(node["pmFwdLADecreaseSize"], m_PMFwdConfig.decreaseSize, m_PMFwdConfig.decreaseSize);
+        readIntegerSize(node["pmMinFwdLASize"], m_PMFwdConfig.minSize, m_PMFwdConfig.minSize);
+        readIntegerSize(node["pmMaxFwdLASize"], m_PMFwdConfig.maxSize, m_PMFwdConfig.maxSize);
+
+        cv::read(node["unwrapRes"], m_UnwrapRes, m_UnwrapRes);
+        cv::read(node["croppedRect"], m_CroppedRect, m_CroppedRect);
+
+        cv::String maskImageFilename;
+        cv::read(node["maskImageFilename"], maskImageFilename, m_MaskImageFilename);
+        m_MaskImageFilename = (std::string)maskImageFilename;
+
+        cv::String watershedMarkerImageFilename;
+        cv::read(node["watershedMarkerImageFilename"], watershedMarkerImageFilename, m_WatershedMarkerImageFilename);
+        m_WatershedMarkerImageFilename = (std::string)watershedMarkerImageFilename;
+
+        cv::read(node["joystickDeadzone"], m_JoystickDeadzone, m_JoystickDeadzone);
+        cv::read(node["serverListenPort"], m_ServerListenPort, m_ServerListenPort);
+        cv::read(node["snapshotServerListenPort"], m_SnapshotServerListenPort, m_SnapshotServerListenPort);
+        cv::read(node["bestSnapshotServerListenPort"], m_BestSnapshotServerListenPort, m_BestSnapshotServerListenPort);
+        cv::read(node["autoTrain"], m_AutoTrain, m_AutoTrain);
+        cv::read(node["moveSpeed"], m_MoveSpeed, m_MoveSpeed);
+
+        double trainInterval;
+        cv::read(node["trainInterval"], trainInterval, m_TrainInterval.count());
+        m_TrainInterval = (Milliseconds)trainInterval;
+
+        int frameSkip = m_SkipFrames;
+        cv::read(node["frameSkip"], frameSkip, frameSkip);
```

Comment:
might be worth making clear exactly when these frames are skipped i.e ``m_TestSkipFrames``

Reply:
:+1: